CMakeLists.txt

```diff
@@ -1217,6 +1217,14 @@ if(NOT CONFIG_BUILD_NO_GAP_FILL)
   set(GAP_FILL "--gap-fill;0xff")
 endif()
 
+if(CONFIG_OUTPUT_PRINT_MEMORY_USAGE)
+  # @Intent: Use the toolchain bintools method for printing memory usage
+  bintools_print_mem_usage(
+    RESULT_CMD_LIST    post_build_commands
+    RESULT_BYPROD_LIST post_build_byproducts
```

Comment:
This API doesn't make sense. Does the implementation append to post_build_commands?

If so it is cleaner to have the implementation set an output variable, and then append those output variables to the post_build_commands from up here.

Reply:
Yes, it is able to add to the post build commands, but it is optional, which is especially clear here since with gnu the bintools_print_mem_usage does not need to add anything to post_build_commands or to byproducts, but another toolchain/bintools most likely will.

In essence, the implementation is setting variables, it is simply placing the results directly where they are needed, instead of first having to add to new variables and then adding those variables to the build commands and byproducts. Having the results placed directly at the final destination, makes it a lot clearer what the implementation is doing and also provides of a lot cleaner and simpler CMakeLists.txt file. 
I sincerely hope that we can keep this interface, as it make the CMakeLists.txt file less cluttered and eliminates extra and unnecessary variables. This interface will be the same for the upcoming abstractions for the remainder of the binary tools used in top level CMakeLists.txt. 
Examples: objcopy
- Usage in CONFIG_GEN_ISR_TABLES
  set(objCpyCmd "")
  bintools_objcopy(
    RESULT_CMD_LIST objCpyCmd
    TARGET_INPUT       ${OUTPUT_FORMAT}
    TARGET_OUTPUT   "binary"
    SECTION_ONLY      ".intList"
    FILE_INPUT              $<TARGET_FILE:${ZEPHYR_PREBUILT_EXECUTABLE}>
    FILE_OUTPUT         "isrList.bin"
    )
  add_custom_command(
    OUTPUT isr_tables.c
    ${objCpyCmd}
    COMMAND ${PYTHON_EXECUTABLE}
...
- Usage in CONFIG_BUILD_OUTPUT_HEX
if(CONFIG_BUILD_OUTPUT_HEX)
  bintools_objcopy(
    RESULT_CMD_LIST       post_build_commands
    RESULT_BYPROD_LIST post_build_byproducts
    STRIP_ALL
    GAP_FILL                    ${GAP_FILL}
    TARGET_OUTPUT      "ihex"
    SECTION_REMOVE     ".comment;COMMON;.eh_frame"
    FILE_INPUT                  ${KERNEL_ELF_NAME}
    FILE_OUTPUT              ${KERNEL_HEX_NAME}
    )
endif()

What do you think?

Reply:
This is not OK. Functions should be given just as much information as they need to do their job, and no more.

In this case we are leaking information about what the other 'post_build_commands' are into the function.

When I read a function like this I assume that the function uses this information (or else we wouldn't have given it to the function in the first place). This assumption is wrong in this case and this makes the function difficult to understand.

When using an output variable that is set by the function it is self-evident from the usage what the function is doing.

Reply:
So, instead of:
  bintools_print_mem_usage(
    RESULT_CMD_LIST    post_build_commands
    RESULT_BYPROD_LIST post_build_byproducts
  )
you want it to be:
  set(memUsageCmd "")
  set(memUsageByProd "")
  bintools_print_mem_usage(
    RESULT_CMD_LIST memUsageCmd
    RESULT_BYPROD_LIST memUsageByProd
    )
  list(APPEND 
    post_build_commands
    ${memUsageCmd}
    )
  list(APPEND
    post_build_byproducts
    ${memUsageByProd}
    )
is that correct? Or do you have another suggestion?

As the parameter states, they are RESULTs, which indicates that they are places for function return information/values and thus I would think it unlikely that anyone would expect the function to look for information in those, and not just add its result as needed.

I do not see how the above is clearer or easier to understand than my original function call, but at least to me, it looks a lot more complicated.

Reply:
Yes.

Reply:
Done.